mars/learn/linear_model/_base.py

```diff
@@ -0,0 +1,418 @@
+# Copyright 1999-2021 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABCMeta, abstractmethod
+import numbers
+
+from sklearn.utils.validation import (check_is_fitted,
+                                      _deprecate_positional_args)
+from sklearn.utils.sparsefuncs import mean_variance_axis
+from sklearn.utils.extmath import safe_sparse_dot
+from sklearn.utils.fixes import delayed
+
+from numpy.linalg import LinAlgError
+from joblib import Parallel
+import scipy.sparse as sp
+from scipy import sparse
+from scipy import optimize
+from scipy.sparse.linalg import lsqr as sparse_lsqr
+from scipy import linalg
+
+from ..base import BaseEstimator, RegressorMixin, MultiOutputMixin
+from ..utils.validation import _check_sample_weight, check_array, FLOAT_DTYPES
+from ..utils.sparsefuncs import inplace_column_scale
+from ..preprocessing import normalize as f_normalize
+import mars.tensor as mt
+from mars.tensor.datasource import tensor as astensor
+
+
+def _preprocess_data(
+    X,
+    y,
+    fit_intercept,
+    normalize=False,
+    copy=True,
+    sample_weight=None,
+    return_mean=False,
+    check_input=True,
+):
+    """Center and scale data.
+
+    Centers data to have mean zero along axis 0. If fit_intercept=False or if
+    the X is a sparse matrix, no centering is done, but normalization can still
+    be applied. The function returns the statistics necessary to reconstruct
+    the input data, which are X_offset, y_offset, X_scale, such that the output
+
+        X = (X - X_offset) / X_scale
+
+    X_scale is the L2 norm of X - X_offset. If sample_weight is not None,
+    then the weighted mean of X and y is zero, and not the mean itself. If
+    return_mean=True, the mean, eventually weighted, is returned, independently
+    of whether X was centered (option used for optimization with sparse data in
+    coordinate_descend).
+
+    This is here because nearly all linear models will want their data to be
+    centered. This function also systematically makes y consistent with X.dtype
+    """
+    if isinstance(sample_weight, numbers.Number):
+        sample_weight = None
+    if sample_weight is not None:
+        sample_weight = astensor(sample_weight)
+
+    if check_input:
+        X = check_array(X, copy=copy, accept_sparse=['csr', 'csc'],
```

Comment:
Can `accept_sparse` be a list of string?

Reply:
accept_sparse could only be True for Mars.